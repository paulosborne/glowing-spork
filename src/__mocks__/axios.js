export const mockPost = jest.fn();
export const mockGet = jest
  .fn()
  .mockImplementation(() => Promise.resolve({ data: undefined }));
export const mockRequest = jest.fn();
export const mockPut = jest.fn();
export const mockRequestInterceptor = jest.fn();

export const mockCreate = jest.fn().mockImplementation(() => ({
  get: mockGet,
  post: mockPost,
  put: mockPut,
  interceptors: {
    request: {
      use: mockRequestInterceptor,
    },
  },
}));

module.exports = {
  mockRequestInterceptor,
  create: mockCreate,
  post: mockPost,
  put: mockPut,
  get: mockGet,
};
