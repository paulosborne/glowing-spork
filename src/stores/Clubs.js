import { flow, types } from 'mobx-state-tree';
import isEmpty from 'lodash.isempty';
import Club from './entities/Club';
import FootballData from '../services/FootballData';

const Clubs = types
  .model('Clubs', {
    data: types.optional(types.array(Club), []),
    selected: types.maybeNull(types.reference(Club)),
  })
  .views(self => ({
    get sortedByName() {
      return self.data
        .slice()
        .sort((a, b) => (a.name.charCodeAt(0) > b.name.charCodeAt(0) ? 1 : -1));
    },
  }))
  .actions(self => ({
    isValidClub(id) {
      return !isEmpty(self.data.filter(club => club.id === id));
    },
    /**
     * Fetches all teams for the given competition and season
     * @param {number} id
     * @param {season} number
     */
    fetchByCompetitionId: flow(function* (id, season) {
      const { data } = yield FootballData.get(`/competitions/${id}/teams`, {
        params: {
          season,
        },
      });

      // clear selected club if it doesn't exist in the new data
      if (
        self.selected !== null &&
        !data.teams.find(({ id }) => id === self.selected.id)
      ) {
        self.selected = null;
      }

      self.data.replace(data.teams);

      if (self.selected === null) {
        self.selectRandom();
      }
    }),

    /**
     * Sets the currently selected club
     * @param {number} id
     */
    selectClub(id) {
      if (self.isValidClub(id)) {
        self.selected = id;
      } else {
        self.selected = null;
      }
    },

    /**
     * Selects a random club
     */
    selectRandom() {
      const index = Math.floor(Math.random() * 6) + 1;
      const { id } = self.data[index];
      self.selectClub(id);
    },
  }));

export default Clubs;
