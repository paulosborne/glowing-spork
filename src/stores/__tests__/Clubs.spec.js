import { getSnapshot, isStateTreeNode } from 'mobx-state-tree';
import Clubs from '../Clubs';
import axios from 'axios';
import MockCompetitionResponse from 'MockCompetitionResponse';

const { teams } = MockCompetitionResponse;
const SEASON = 2020;
const COMPETITION_ID = 2021;

describe('Store: Clubs', () => {
  let instance;

  beforeEach(() => {
    instance = Clubs.create();
  });

  it('should create the store correctly', () => {
    expect(isStateTreeNode(instance)).toEqual(true);
    expect(getSnapshot(instance)).toMatchSnapshot();
  });

  /**
   * fetchByCompetitionId()
   */
  describe('fetchByCompetitionId()', () => {
    it('should be defined', () => {
      expect(instance.fetchByCompetitionId).toBeDefined();
    });

    it('should call the football data API correctly', async () => {
      axios.get.mockResolvedValue({ data: { teams } });
      await instance.fetchByCompetitionId(COMPETITION_ID, SEASON);
      expect(axios.get).toHaveBeenCalledWith(
        `/competitions/${COMPETITION_ID}/teams`,
        {
          params: { season: SEASON },
        },
      );
      expect(instance.data.length).toEqual(teams.length);
    });
  });
});
