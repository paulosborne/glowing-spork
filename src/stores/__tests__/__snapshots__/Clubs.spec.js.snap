// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store: Clubs should create the store correctly 1`] = `
Object {
  "data": Array [],
  "selected": null,
}
`;
